File: auntieos-admin/android/app/src/test/java/com/tribetails/auntieos/ui/admin/ActivityDetailTest.kt
package com.tribetails.auntieos.ui.admin

import com.tribetails.auntieos.data.admin.ActivityLogEntry
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The Activity Log's full sealed record, Android side.
 *
 * Operator ruling R5, 2026-08-03: "the Activity Log is seriously lacking, cant
 * see shit or what the fuck actually happened."
 *
 * Android was ahead of web here. Rows were already clickable and a detail
 * overlay already existed, and it STILL could not answer that, because the
 * overlay showed six fields and none of them was `payload`. Every field asserted
 * below has been sealed onto every entry by `writeAuditEntry` since 2026-05-19;
 * its own docstring calls them "retained for forensic value ... surfaced in
 * detail views", and they were surfaced in none.
 */
class ActivityDetailTest {

    private fun entry(
        payload: Map<String, Any?> = emptyMap(),
        seq: Long? = 1482L,
        entryHash: String = "",
        prevHash: String = "",
        severity: String = "",
        actorRole: String = "",
        targetCollection: String = "",
        targetId: String = "",
        ip: String = "",
    ) = ActivityLogEntry(
        id = "a1",
        timestamp = "2026-08-03T14:02:11.482Z",
        actionType = "NOTIFICATION_RECEIVED",
        description = "Notification kincare.booking.confirm delivered via email",
        status = "SUCCESS",
        seq = seq,
        entryHash = entryHash,
        prevHash = prevHash,
        payload = payload,
        severity = severity,
        actorRole = actorRole,
        targetCollection = targetCollection,
        targetId = targetId,
        ip = ip,
    )

    @Test
    fun detailLeadsWithTheFullTimestampThatTheRowTruncates() {
        // The row shows `HH:mm`. The point of opening an entry is to see what it
        // cut, and "which second" is routinely the question when reconciling
        // against a provider's logs.
        assertEquals("When" to "2026-08-03T14:02:11.482Z", activityDetailRows(entry()).first())
    }

    @Test
    fun detailSurfacesTheProvenanceFieldsTheOverlayNeverShowed() {
        val labels = activityDetailRows(
            entry(severity = "info", actorRole = "SYSTEM", ip = "203.0.113.7"),
        ).map { it.first }
        assertTrue(labels.contains("Severity"))
        assertTrue(labels.contains("Actor role"))
        assertTrue(labels.contains("IP"))
    }

    @Test
    fun detailRendersTheTargetAsAPathNeverAFabricatedLink() {
        val rows = activityDetailRows(entry(targetCollection = "notifications", targetId = "n1"))
        assertTrue(rows.contains("Target" to "notifications/n1"))
    }

    @Test
    fun detailFallsBackToAGenericSegmentRatherThanDroppingABareTargetId() {
        assertTrue(activityDetailRows(entry(targetId = "n1")).contains("Target" to "target/n1"))
    }

    @Test
    fun detailOmitsFieldsTheWriterDidNotRecord() {
        val labels = activityDetailRows(entry()).map { it.first }
        assertTrue(!labels.contains("IP"))
        assertTrue(!labels.contains("Target"))
    }

    /**
     * THE FIELD THAT ANSWERS THE COMPLAINT. The description says a notification
     * was delivered via email; the payload says to whom, for which notification,
     * and with which provider message id.
     */
    @Test
    fun payloadFlattensWhatActuallyHappened() {
        val rows = activityPayloadRows(
            entry(
                payload = mapOf(
                    "notificationId" to "n1",
                    "key" to "kincare.booking.confirm",
                    "channel" to "email",
                    "providerMessageId" to "sg-88",
                ),
            ),
        )
        assertEquals(
            listOf(
                "channel" to "email",
                "key" to "kincare.booking.confirm",
                "notificationId" to "n1",
                "providerMessageId" to "sg-88",
            ),
            rows,
        )
    }

    @Test
    fun payloadFlattensNestedMapsToDottedPaths() {
        assertEquals(
            listOf("result.scanned" to "40", "result.split" to "3"),
            activityPayloadRows(entry(payload = mapOf("result" to mapOf("split" to 3, "scanned" to 40)))),
        )
    }

    @Test
    fun payloadKeepsAnArrayOnOneLineBecauseItIsOneField() {
        assertEquals(
            listOf("channels" to "email, sms"),
            activityPayloadRows(entry(payload = mapOf("channels" to listOf("email", "sms")))),
        )
    }

    /**
     * A recorded null is a decision the writer made; blanking it hides that, and
     * in an audit trail "the writer recorded null here" is evidence.
     */
    @Test
    fun payloadPrintsARecordedNullAndAnEmptyMap() {
        assertEquals(
            listOf("providerMessageId" to "null"),
            activityPayloadRows(entry(payload = mapOf("providerMessageId" to null))),
        )
        assertEquals(
            listOf("extra" to "{}"),
            activityPayloadRows(entry(payload = mapOf("extra" to emptyMap<String, Any?>()))),
        )
    }

    @Test
    fun payloadIsEmptyForAnEntryWrittenWithoutOne() {
        assertTrue(activityPayloadRows(entry()).isEmpty())
    }

    /**
     * FULL hashes. The row abbreviates to eight characters because it has one
     * line; an opened entry is where someone verifies a hash by eye against
     * verifyActivityLogChain, and a truncated hash verifies nothing.
     */
    @Test
    fun chainShowsTheSequenceAndTheFullHashes() {
        val full = "a".repeat(64)
        val prev = "b".repeat(64)
        assertEquals(
            listOf("Sequence" to "#1482", "Entry hash" to full, "Previous hash" to prev),
            activityChainRows(entry(entryHash = full, prevHash = prev)),
        )
    }

    @Test
    fun chainIsEmptyForALegacyPreChainEntrySoTheScreenCanSaySo() {
        assertTrue(activityChainRows(entry(seq = null)).isEmpty())
    }

    @Test
    fun chainStillReportsTheSequenceForAnEntryWithNoHashes() {
        assertEquals(listOf("Sequence" to "#1482"), activityChainRows(entry()))
    }
}

File: auntieos-admin/src/lib/activityDetail.test.ts
import { describe, it, expect } from 'vitest';
import type { ActivityLogEntry } from '../api/activityLog';
import {
  activityChainRows,
  activityDetailRows,
  activityMatchesQuery,
  activityMatchesStatus,
  activityPayloadRows,
} from './activityDetail';

function entry(over: Partial<ActivityLogEntry> = {}): ActivityLogEntry {
  return {
    _id: 'a1',
    timestamp: '2026-08-03T14:02:11.482Z',
    actionType: 'NOTIFICATION_RECEIVED',
    description: 'Notification kincare.booking.confirm delivered via email',
    status: 'SUCCESS',
    actorId: '',
    seq: 1482,
    ...over,
  };
}

/**
 * R5 on the Activity Log: "the Activity Log is seriously lacking, cant see shit
 * or what the fuck actually happened."
 *
 * Every field these selectors read has been on the document since 2026-05-19;
 * `writeAuditEntry` calls them "retained for forensic value ... surfaced in
 * detail views" and no detail view existed. So these tests are about DISCLOSURE,
 * not about new data.
 */
describe('activityDetailRows', () => {
  it('leads with the FULL timestamp, which is the part the collapsed row truncates', () => {
    const rows = activityDetailRows(entry());
    expect(rows[0]).toEqual({ label: 'When', value: '2026-08-03T14:02:11.482Z' });
  });

  it('surfaces the provenance fields the row never showed', () => {
    const rows = activityDetailRows(
      entry({
        severity: 'info',
        actorRole: 'SYSTEM',
        familyId: 'fam1',
        requestId: 'req-9',
        ip: '203.0.113.7',
        userAgent: 'Mozilla/5.0',
      }),
    );
    const labels = rows.map((r) => r.label);
    expect(labels).toContain('Severity');
    expect(labels).toContain('Actor role');
    expect(labels).toContain('Household');
    expect(labels).toContain('Request');
    expect(labels).toContain('IP');
    expect(labels).toContain('User agent');
  });

  it('renders the target as a path, never a fabricated link', () => {
    const rows = activityDetailRows(entry({ targetCollection: 'notifications', targetId: 'n1' }));
    expect(rows).toContainEqual({ label: 'Target', value: 'notifications/n1' });
  });

  it('falls back to a generic collection segment rather than dropping a bare target id', () => {
    const rows = activityDetailRows(entry({ targetId: 'n1' }));
    expect(rows).toContainEqual({ label: 'Target', value: 'target/n1' });
  });

  it('omits fields the writer did not record', () => {
    const labels = activityDetailRows(entry()).map((r) => r.label);
    expect(labels).not.toContain('IP');
    expect(labels).not.toContain('Target');
  });
});

describe('activityPayloadRows', () => {
  /**
   * THE FIELD THAT ANSWERS THE COMPLAINT. The description says a notification
   * was delivered via email; the payload says to whom, for which notification,
   * and with which provider message id, and it was rendered nowhere in the
   * product.
   */
  it('flattens the payload that says what actually happened', () => {
    const rows = activityPayloadRows(
      entry({
        payload: {
          notificationId: 'n1',
          key: 'kincare.booking.confirm',
          channel: 'email',
          recipientUid: 'u1',
          providerMessageId: 'sg-88',
        },
      }),
    );
    expect(rows).toEqual([
      { label: 'channel', value: 'email' },
      { label: 'key', value: 'kincare.booking.confirm' },
      { label: 'notificationId', value: 'n1' },
      { label: 'providerMessageId', value: 'sg-88' },
      { label: 'recipientUid', value: 'u1' },
    ]);
  });

  it('sorts keys, so two entries of the same type are comparable by eye', () => {
    expect(activityPayloadRows(entry({ payload: { z: 1, a: 2 } })).map((r) => r.label)).toEqual([
      'a',
      'z',
    ]);
  });

  it('flattens nested maps to dotted paths rather than dumping JSON', () => {
    const rows = activityPayloadRows(entry({ payload: { result: { split: 3, scanned: 40 } } }));
    expect(rows).toEqual([
      { label: 'result.scanned', value: '40' },
      { label: 'result.split', value: '3' },
    ]);
  });

  it('keeps an array on one line, because it is one field', () => {
    expect(activityPayloadRows(entry({ payload: { channels: ['email', 'sms'] } }))).toEqual([
      { label: 'channels', value: 'email, sms' },
    ]);
  });

  /**
   * A recorded null is a decision the writer made. Blanking it would hide that,
   * and in an audit trail "the writer recorded null here" is evidence.
   */
  it('prints a recorded null as null, and an empty map as {}', () => {
    expect(activityPayloadRows(entry({ payload: { providerMessageId: null } }))).toEqual([
      { label: 'providerMessageId', value: 'null' },
    ]);
    expect(activityPayloadRows(entry({ payload: { extra: {} } }))).toEqual([
      { label: 'extra', value: '{}' },
    ]);
  });

  it('is empty for an entry written with no payload, and for a malformed one', () => {
    expect(activityPayloadRows(entry())).toEqual([]);
    expect(activityPayloadRows(entry({ payload: 'nope' }))).toEqual([]);
    expect(activityPayloadRows(entry({ payload: null }))).toEqual([]);
  });
});

describe('activityChainRows', () => {
  /**
   * FULL hashes. The collapsed row abbreviates to eight characters because it
   * has one line; an opened entry is where someone verifies a hash by eye
   * against verifyActivityLogChain, and a truncated hash verifies nothing.
   */
  it('shows the sequence and the full hashes, never a prefix', () => {
    const full = 'a'.repeat(64);
    const prev = 'b'.repeat(64);
    expect(activityChainRows(entry({ entryHash: full, prevHash: prev }))).toEqual([
      { label: 'Sequence', value: '#1482' },
      { label: 'Entry hash', value: full },
      { label: 'Previous hash', value: prev },
    ]);
  });

  it('is empty for a legacy pre-chain entry, so the screen can say so', () => {
    const { seq: _dropped, ...legacy } = entry();
    expect(activityChainRows(legacy)).toEqual([]);
  });

  it('still reports the sequence for an entry that has one but no hashes', () => {
    expect(activityChainRows(entry())).toEqual([{ label: 'Sequence', value: '#1482' }]);
  });
});

describe('activityMatchesQuery', () => {
  it('matches the visible fields', () => {
    expect(activityMatchesQuery(entry(), 'notification_received')).toBe(true);
    expect(activityMatchesQuery(entry(), 'delivered via email')).toBe(true);
  });

  /**
   * The reason search is worth having: "which entry mentions this booking id"
   * is the real operator question, and the id is almost always in the payload
   * rather than in the description.
   */
  it('matches inside the payload, which is where the ids actually are', () => {
    expect(activityMatchesQuery(entry({ payload: { bookingId: 'b_7fJ2' } }), 'b_7fj2')).toBe(true);
  });

  it('is case-insensitive and treats an empty query as no filter', () => {
    expect(activityMatchesQuery(entry(), 'NOTIFICATION')).toBe(true);
    expect(activityMatchesQuery(entry(), '   ')).toBe(true);
  });

  it('answers false rather than falling back to everything on no match', () => {
    expect(activityMatchesQuery(entry(), 'zzzz')).toBe(false);
  });
});

describe('activityMatchesStatus', () => {
  it('folds FAILURE and the legacy ERROR spelling into one Problems bucket', () => {
    expect(activityMatchesStatus(entry({ status: 'FAILURE' }), 'problems')).toBe(true);
    expect(activityMatchesStatus(entry({ status: 'ERROR' }), 'problems')).toBe(true);
    expect(activityMatchesStatus(entry({ status: 'SUCCESS' }), 'problems')).toBe(false);
  });

  it('passes everything through on All', () => {
    expect(activityMatchesStatus(entry({ status: '' }), 'all')).toBe(true);
  });

  it('matches success and pending exactly, whatever the casing on the wire', () => {
    expect(activityMatchesStatus(entry({ status: ' success ' }), 'success')).toBe(true);
    expect(activityMatchesStatus(entry({ status: 'PENDING' }), 'pending')).toBe(true);
    expect(activityMatchesStatus(entry({ status: 'PENDING' }), 'success')).toBe(false);
  });

  it('does not bucket a row with no status anywhere but All', () => {
    const { status: _dropped, ...noStatus } = entry();
    expect(activityMatchesStatus(noStatus, 'all')).toBe(true);
    expect(activityMatchesStatus(noStatus, 'success')).toBe(false);
    expect(activityMatchesStatus(noStatus, 'problems')).toBe(false);
  });
});
